scale-out-relay/src/main/scala/com/normation/plugins/scaleoutrelay/ScaleOutRelayService.scala

```diff
@@ -39,119 +41,87 @@ class ScaleOutRelayService(
-//      _ <- if(level >= 5) woDirectiveRepository.deleteActiveTechnique(activeTechniqueIdCommon, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed active technique ${activeTechniqueIdCommon.value} failed")
-//      _ <- if(level >= 6) woRuleRepository.delete(ruleSetup.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed rule  ${ruleSetup.id.value} failed")
-//      _ <- if(level >= 7) woRuleRepository.delete(ruleDistribPolicy.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed rule  ${ruleDistribPolicy.id.value} failed")
-//    } yield ()
-//  }
+  private[scaleoutrelay] def rollbackPromote(
+      ruleTarget              : PolicyServerTarget
+    , ruleSetup               : Rule
+    , ruleDistribPolicy       : Rule
+    , newInfo                 : NodeInfo
+    , nodeGroup               : NodeGroup
+    , activeTechniqueId       : ActiveTechniqueId
+    , activeTechniqueIdCommon : ActiveTechniqueId
+    , modId                   : ModificationId
+    , actor                   : EventActor
+  ) : List[ZIO[Any, Object, Any]] = {
+
+    val old = PolicyServerToNode(newInfo)
+    val reason = Some("Promote node to relay rollback")
+
+    val a = woLDAPNodeRepository.createNode(old.node, modId, actor, reason)
+             .chainError(s"Rollback promote to relay failed : restore ${newInfo.node} configuration failed")
+    val b  = woLDAPNodeGroupRepository.deletePolicyServerTarget(ruleTarget)
+             .chainError(s"Rollback promote to relay failed : removing ${newInfo.node} configuration failed")
+    val c = woLDAPNodeGroupRepository.delete(nodeGroup.id, modId, actor, reason)
+             .chainError(s"Rollback promote to relay failed : removing node group ${nodeGroup.id.value} failed")
+    val d = woDirectiveRepository.deleteActiveTechnique(activeTechniqueId, modId, actor, reason)
+             .chainError(s"Rollback promote to relay failed : removing active technique ${activeTechniqueId.value} failed")
+    val e = woDirectiveRepository.deleteActiveTechnique(activeTechniqueIdCommon, modId, actor, reason)
+             .chainError(s"Rollback promote to relay failed : removing active technique ${activeTechniqueIdCommon.value} failed")
+
+    // Rules deletion return an error if there is no such ID
+    val f = woRuleRepository.delete(ruleSetup.id, modId, actor, reason).catchAll {
+      err =>
+        ZIO.effect(PluginLogger.info(s"Trying to remove residual object Rule ${ruleSetup.id} : ${err.fullMsg}"))
+    }
+    val g = woRuleRepository.delete(ruleDistribPolicy.id, modId, actor, reason).catchAll {
+      err =>
+        ZIO.effect(PluginLogger.info(s"Trying to remove residual object Rule ${ruleDistribPolicy.id} : ${err.fullMsg}"))
+    }
+
+    List(a,b,c,d,e,f,g)
```

Comment:
What I want to do here it's to regroup all `ZIO[Any, RudderError, Something]` in a List to return it. Then when I need to perform a rollback I collect the element of the list and accumulate it to return a `ZIO[Any, RudderError, NodeInfo]`

Reply:
`ZIO.collectAll`

Reply:
sorry, you need an either somewher to avoid failing on first error with collectAll. But we have `.accumulate` that does exactly that.

Reply:
I don't think you want to return nodInfo, the return type is not important. If you really want, you need to fold them into your result but I don't think that's what you want

You wanted the nodeInfo because you used a catchall ?

Reply:
ie `List(a,b,c,d,e,f,g).accumulate` as long you import com.normation.errors._

Reply:
You are right, the returned type doesn't matter